### PR TITLE
GFORMS-1668 - Add progressIndicatorSize.

### PR DESCRIPTION
**Pull request by rhubner**

### CODE REVIEW

**Review by josef-vlach — CHANGES_REQUESTED** [review]
Can we have a unit tests for these transformation please?

File: app/uk/gov/hmrc/gform/formtemplate/Substituter.scala [hunk 1]
```diff
@@ -197,7 +205,7 @@ object Substituter {
     t.copy(
       title = t.title(substitutions),
       noPIITitle = t.noPIITitle(substitutions),
-      description = t.description(substitutions),
+      description = t.description.apply(substitutions),
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/formtemplate/Substituter.scala:208]: Why this change?

Comment (rhubner) [reply app/uk/gov/hmrc/gform/formtemplate/Substituter.scala:208]: No idea. Probably typo.

File: app/uk/gov/hmrc/gform/formtemplate/FormTemplatesControllerRequestHandler.scala [hunk 2]
```diff
@@ -164,6 +164,48 @@ object FormTemplatesControllerRequestHandler {
...
+        JsArray(
+          pages.value.map {
+            case page: JsObject if page.value.contains("progressIndicator") =>
+              transformProgressIndicator(page)
+            case x => x
+          }
+        )
+      }
+
+    def transformProgressIndicator(data: JsObject) = {
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/formtemplate/FormTemplatesControllerRequestHandler.scala:178]: Would it be possible to write this in transformers style, i.e. that it returns `Reads[JsObject]` and you know that this act on section json object?

File: app/uk/gov/hmrc/gform/formtemplate/FormTemplatesControllerRequestHandler.scala [hunk 3]
```diff
@@ -164,6 +164,48 @@ object FormTemplatesControllerRequestHandler {
         Json.obj()
       )
 
+    def transformProgressIndicatiorInArray =
+      __.read[JsArray].map { pages =>
+        JsArray(
+          pages.value.map {
+            case page: JsObject if page.value.contains("progressIndicator") =>
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/formtemplate/FormTemplatesControllerRequestHandler.scala:171]: Why is this condition needed. Either you can do the transformation or you can't. 

Writing `transformProgressIndicator` in transformer style would allow to `transformProgressIndicatiorInArray` more concisely in terms `Reads` api, rather than this mix of style right now.

**Review by josef-vlach — CHANGES_REQUESTED** [review]

File: test/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormTemplateJSONSpec.scala [hunk 4]
```diff
@@ -424,12 +424,319 @@ class FormTemplateJSONSpec extends Spec with TableDrivenPropertyChecks {
...
+          |            "infoText": "You must tell us about all directors"
+          |          }
+          |        ]
+          |      },
+          |      "pages": [
+          |        {
+          |          "progressIndicator": {
+          |            "label": "Name",
+          |            "labelSize": {
+          |              "Large": {}
```
[the remark this answers is not in the document]
Comment (rhubner) [reply test/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormTemplateJSONSpec.scala:628]: Thats how `LabelSize` is serialized. https://github.com/hmrc/gform/blob/main/app/uk/gov/hmrc/gform/sharedmodel/formtemplate/LabelSize.scala#L31

File: test/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormTemplateJSONSpec.scala [hunk 5]
```diff
@@ -424,12 +424,319 @@ class FormTemplateJSONSpec extends Spec with TableDrivenPropertyChecks {
...
+          |            }
+          |          },
+          |          "progressIndicatorSize": "l",
```
[the remark this answers is not in the document]
Comment (rhubner) [reply test/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormTemplateJSONSpec.scala]: done

File: test/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormTemplateJSONSpec.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (josef-vlach) [reply test/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormTemplateJSONSpec.scala:628]: So do not use `LabelSize` and keep it as `String`

File: app/uk/gov/hmrc/gform/formtemplate/FormTemplatesControllerRequestHandler.scala [hunk 6]
```diff
@@ -164,6 +164,48 @@ object FormTemplatesControllerRequestHandler {
...
+      __.json.update(createProgressIndicator) andThen (__ \ 'progressIndicatorSize).json.prune orElse (__.read)
+
+    }
+
+    val transformProgressIndicatiorInArray =
+      __.read[JsArray].map { pages =>
+        JsArray(
+          pages.value.map(_.transform(transformProgressIndicator).get)
+        )
+      }
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/formtemplate/FormTemplatesControllerRequestHandler.scala:191]: This can be replaced by:

```suggestion
    val transformProgressIndicatiorInArray: Reads[JsArray] = Reads.list(transformProgressIndicator).map(JsArray.apply)
```

**Review by josef-vlach — APPROVED** [review]
